onnxruntime/core/providers/cuda/cuda_utils.cu

```diff
@@ -56,9 +56,27 @@ std::unique_ptr<IConstantBuffer<T>> CreateConstantOnes() {
   return std::make_unique<ConstantBufferImpl<T>>(Consts<T>::One);
 }
 
+template <typename T>
+std::unique_ptr<IConstantBuffer<T>> CreateConstantValue(T value) {
+  return std::make_unique<ConstantBufferImpl<T>>(value);
+}
+
 template std::unique_ptr<IConstantBuffer<float>> CreateConstantOnes<float>();
 template std::unique_ptr<IConstantBuffer<double>> CreateConstantOnes<double>();
 template std::unique_ptr<IConstantBuffer<half>> CreateConstantOnes<half>();
 
+template std::unique_ptr<IConstantBuffer<float>> CreateConstantValue<float>(float value);
+template std::unique_ptr<IConstantBuffer<double>> CreateConstantValue<double>(double value);
+template std::unique_ptr<IConstantBuffer<half>> CreateConstantValue<half>(half value);
+template std::unique_ptr<IConstantBuffer<bool>> CreateConstantValue<bool>(bool value);
+template std::unique_ptr<IConstantBuffer<int8_t>> CreateConstantValue<int8_t>(int8_t value);
+template std::unique_ptr<IConstantBuffer<int16_t>> CreateConstantValue<int16_t>(int16_t value);
+template std::unique_ptr<IConstantBuffer<int32_t>> CreateConstantValue<int32_t>(int32_t value);
+template std::unique_ptr<IConstantBuffer<int64_t>> CreateConstantValue<int64_t>(int64_t value);
+template std::unique_ptr<IConstantBuffer<uint8_t>> CreateConstantValue<uint8_t>(uint8_t value);
+template std::unique_ptr<IConstantBuffer<uint16_t>> CreateConstantValue<uint16_t>(uint16_t value);
+template std::unique_ptr<IConstantBuffer<uint32_t>> CreateConstantValue<uint32_t>(uint32_t value);
+template std::unique_ptr<IConstantBuffer<uint64_t>> CreateConstantValue<uint64_t>(uint64_t value);
```

Comment:
I'd recommend not to use IConstantBuffer to fill constant values, as the buffer is not going to be released. You may reuse the CUDA kernel underneath though.

Reply:
Keep it will bring the benefit to reuse it for next run if the shape is the same. But the bad part is there could be many ConstantOfShape with large shape.
But release the memory and always create new also has some problem. We need to make sure the Async memcpy completed before release the memory.

---
In reply to: [290942884](https://github.com/microsoft/onnxruntime/pull/1168#discussion_r290942884) [](ancestors = 290942884)

Reply:
Copy is slower than fill as it does both read and write. Besides, for ConstantOfShape you only need to fill uint8/uint16/uint32/uint64, and do the case in CPU.

---
In reply to: [290951162](https://github.com/microsoft/onnxruntime/pull/1168#discussion_r290951162) [](ancestors = 290951162,290942884)

Reply:
agree. maybe you can try [https://thrust.github.io/doc/group__filling.html](https://thrust.github.io/doc/group__filling.html) if it works

---
In reply to: [290952977](https://github.com/microsoft/onnxruntime/pull/1168#discussion_r290952977) [](ancestors = 290952977,290951162,290942884)